torchaudio/prototype/ctc_decoder/ctc_decoder.py

```diff
@@ -0,0 +1,224 @@
+import torch
+import itertools as it
+from typing import List, Optional, Dict
+from collections import namedtuple
+
+import torchaudio
+
+try:
+    torchaudio._extension._load_lib('libtorchaudio_decoder')
+    from torchaudio._torchaudio_decoder import (
+        _CriterionType,
+        _KenLM,
+        _LexiconDecoder,
+        _LexiconDecoderOptions,
+        _SmearingMode,
+        _Trie,
+        _Dictionary,
+        _create_word_dict,
+        _load_words,
+    )
+except ImportError:
+    raise ImportError(
+        "flashlight decoder bindings are required to use this functionality. "
+        "Please set BUILD_CTC_DECODER=1 when building from source."
+    )
+
+__all__ = ["KenLMLexiconDecoder", "kenlm_lexicon_decoder"]
+
+
+Hypothesis = namedtuple("Hypothesis", ["tokens", "words", "score"])
+
+class KenLMLexiconDecoder:
+    def __init__(
+        self,
+        nbest: int,
+        lexicon: dict,
+        word_dict: _Dictionary,
+        tokens_dict: _Dictionary,
+        kenlm: _KenLM,
+        decoder_options: _LexiconDecoderOptions,
+        blank_token: str,
+        sil_token: str,
+    ) -> None:
+        """
+        KenLM CTC Decoder with Lexicon constraint.
+
+        Note:
+            To build the decoder, please use the factory function kenlm_lexicon_decoder.
+
+        Args:
+            nbest (int): number of best decodings to return
+            lexicon (dict): lexicon mapping of words to spellings
+            word_dict (_Dictionary): dictionary of words
+            tokens_dict (_Dictionary): dictionary of tokens
+            kenlm (_KenLM): n-gram KenLM language model
+            decoder_options (_LexiconDecoderOptions): parameters used for beam search decoding
+            blank_token (str): token corresopnding to blank
+            sil_token (str): token corresponding to silence
+        """
+
+        self.nbest = nbest
+        self.decoder_options = decoder_options
+        self.lexicon = lexicon
+        self.word_dict = word_dict
+        self.lm = kenlm
+        self.tokens_dict = tokens_dict
+
+        self.vocab_size = tokens_dict.index_size()
+
+        self.unk_word = word_dict.get_index("<unk>")
```

Comment:
I think this unk token should be parameterized as well. `UNK` is another popular one.